Preferences KDD (#5729)
msupply-foundation/open-msupply: decisions/2024-12-03_preferences.md

- We store the pref data in an JSON text field. We deserialize this with serde into structs for each pref as expected.
  - This does mean that sync does not validate the `data` field, that validation occurs when the record is accessed. Though, we could quickly deserialize at sync buffer integration time just to check!
- We can drop the `store_preference` table, but keep the repo the same for now, just querying from `preference` instead.

_Pros:_

- Consistency with OG can be easier to reason about between systems (provided it the OG pattern isn't hard to reason about to begin with...)
- Just 1 table and logic (i.e. 1 repo etc.)
- Generally flexible and battle tested

_Cons:_

- Potentially more upfront effort right now
- Create need for design convention on how to structure preferences. Should you make an individual pref for every little thing, or group them together into many keys in the same pref's JSON? How many is too many? If ever concurrent editing of same record with race on saving, bigger JSON means more data lost without merge strategy. We can nest - how much nesting is TOO FAR?

>   - structs defining pref shape should at least provide some tacit guardrails; if your struct is crazy maybe the pref should be broken down!

Reviewer: ```suggestion
  - structs defining pref shape should at least provide some tacit guardrails; if your struct is crazy maybe the pref should be broken down!
   - Requires dedicated code for configuration UI (as we don't know how pref might be structured)
```

### Option 2 - Add more preference tables like the current OMS `store_preference`

_Pros:_

- Lower effort for now

_Cons:_

> - Some preferences might need more than one FK, e.g. user pref for a specific store or machine.

Reviewer: ```suggestion
- Some preferences might need more than one FK, e.g. user pref for a specific store or machine.

### Option 3 - Similar to `properties` and `name_properties` structure

With this option we'd define a list of `prefs` in the database that could be assigned, and what possible values they have e.g. float, int, boolean, controlled list.

Each preference has a `unique_key` assigned, this key is used to assign the value to when serialising/deserialising to json.

Each pref would record if it's relevent to apply at a global, store, or other level.

In sync we'd sync a `preference_data` record for each store and one global (for now) could add user, or machine etc as needed in future. This would have a json payload similar to Option 1.

There's nothing to stop us still deserialise them in a rust struct as needed

This option would allow us to develop a preference editor UI similar to what we have with editing store/name properties.

If you need some more complex type in your sync record, then it probably is worth creating a dedicated ui and management for this struct.

_Pros:_

- Re-uses an existing pattern
- Clear path for creating a UI without lots of over head for adding a new pref
- Can still deserialise to specific structs if needed

_Cons:_
- Could be larger payload as prefs aren't broken down into different areas
- Need to insert prefs into database rather than just adding to code? Maybe there's a code first way to achieve the same thing? Enums?

```

![image](https://github.com/user-attachments/assets/6d1e0653-297d-4e00-a04a-7cdefece9fd6)

Each preference has a `unique_key` assigned, this key is used to assign the value to when serialising/deserialising to json.

Each pref would record if it's relevent to apply at a global, store, or other level.

In sync we'd sync a `preference_data` record for each store and one global (for now) could add user, or machine etc as needed in future. This would have a json payload similar to Option 1.

There's nothing to stop us still deserialise them in a rust struct as needed

This option would allow us to develop a preference editor UI similar to what we have with editing store/name properties.

If you need some more complex type in your sync record, then it probably is worth creating a dedicated ui and management for this struct.

_Pros:_

- Re-uses an existing pattern
- Clear path for creating a UI without lots of over head for adding a new pref
- Can still deserialise to specific structs if needed
- Could be used by plugins to add additional prefs as needed

_Cons:_
- Could be larger payload as prefs aren't broken down into different areas
- Need to insert prefs into database rather than just adding to code? Maybe there's a code first way to achieve the same thing? Enums?

> 

Author: Great suggestion @jmbrunskill !

Would work quite well for plugins as they'd just be centrally controlled records that sync everywhere. 

I'd be slightly less concerned about the larger payload. We have already the "custom_data" field and that hasn't blown up too badly despite the years. In some ways it's quicker to sync 1 big record over 1000 pref records, far few queries. Tricky if someone puts something chunky in it though, such as a base64 encoded file. e.g. a plugin for adding an album of photos of a store, but rather than using document sync they just encode it in the plugin config 😆.

Speaking of plug config, I had been thinking that if plugins had config then the plugin should include the UI for the configuration that the FE loads up. That might have some flexibility advantages, but the way you suggest would probably cover most cases and be easier than cooking a UI for every plugin.

I'll add to the other options details on what config options they have.

#### OMS

We currently just have a basic pref setup for store prefs. There is a `store_preference` table that has a column for each preference. This provides typing. There are some column limits of 1600 for postgres and ~32000 for sqlite, we're unlikely to hit those though.

This doesn't cover things such as machine or user preferences by it self. To follow this pattern for other requirements, you'd have to make the tables `user_preference` and `machine_preference` and so on. If a future requirement, cross overs of those prefs might not really work without getting crazy e.g. a user pref for a particular store.

## Options

### Option 1 - Take the lessons from OG `pref` table by doing the same

The OG pref table is battle tested for over a decade:

- We create a `preference` (or `pref`?) table with `id`, `item`, `store_id`, `user_id`, `machine_id` and `data` fields.

> - The lack of typing that ailed OG can readily be addressed with with structs in OMS. These structs may carry through to the frontend gql types as appropriate.

Reviewer: ```suggestion
- The lack of typing that ailed OG can readily be addressed with a big struct in OMS. Thus structs may carry through to the frontend gql types as appropriate.
```

- Sync needs deeper thought, but at a glance: store prefs are easily identified and synced by the `store_id`. `user_id` could be added to `change_log` to handle that (if needed to sync... or we sync users and their prefs everywhere 😉). Perhaps it's OK to not sync and lose machine_id records.

> - We store the pref data in an JSON text field. We deserialize this with serde into structs for each pref as expected.

Reviewer: ```suggestion
- We store the pref data in an JSON text field. We deserialize this into a struct in the rust code to which provides sensible defaults for anything missing.
```

  - This does mean that sync does not validate the `data` field, that validation occurs when the record is accessed. Though, we could quickly deserialize at sync buffer integration time just to check!
- We can drop the `store_preference` table, but keep the repo the same for now, just querying from `preference` instead.

> 

Reviewer: ```suggestion
- The list of preferences is defined by a single struct in the backend. The backend is responsible for merging Global, Store, and potentially if required `User Prefs/Machine Prefs`.
The single struct can have additional complexity, e.g. struct within structs as required. This may require additional mapping to graphql though.

```

_Pros:_

- Consistency with OG can be easier to reason about between systems (provided it the OG pattern isn't hard to reason about to begin with...)
- Just 1 table and logic (i.e. 1 repo etc.)

> - Generally flexible and battle tested

Reviewer: ```suggestion
- Generally flexible and battle tested
-  Preferences rust struct gives single source of the all preferences defined in the system
```

_Cons:_

- Potentially more upfront effort right now

> - Create need for design convention on how to structure preferences. Should you make an individual pref for every little thing, or group them together into many keys in the same pref's JSON? How many is too many? If ever concurrent editing of same record with race on saving, bigger JSON means more data lost without merge strategy. We can nest - how much nesting is TOO FAR?

Reviewer: I'm not entirely sure if I still agree with this comment, given it's one struct we don't have different `prefs` there's just one set of prefs for the whole system (in my understanding?)

Author: I'm not sure what you mean by having different prefs, but having a set (of different?) prefs? If it helps clarify there was a bit of implementation discussion that came up - basically I thought we were heading towards a bigger json pref that was all of a store's prefs, but was worried that we'd be pinned for potential sync situations, leading to clarifying another thing...

In OG store prefs and global prefs are both controlled on the central server, we're not worried about merging pref records through sync atm. That's likely to continue to be the case in OMS, though it'd be nice to keep the door open for some store prefs to be controlled on the remote site. At any rate, if prefs are being edited on both remote and central for a given store, individual records for prefs makes conflict resolution easier. Unless we use CRDTs 😁.

Anyway, so assuming that there are lots of individual prefs, to initialise the BIG STRUCT we'd do something like query all the prefs for a given store_ID then have fns on the struct for getting a particular pref from that selection or default. 

Does that make sense, or does that raise more questions? In fairness if we're all on the same page then this _con_ can be removed because the convention is taking shape already!

Reviewer: Does my rust code below help clarify the `BIG STRUCT` approach?
I was thinking for the first implementation at least, we'd have a single database row per store, and a single global row. The internal JSON structure would be the same, although we wouldn't need to serialise the `None` fields.

If we want to allow remote sites to edit their own store_prefs, then I think we go with the last update wins initially.
Could alway extend to allow some kind of CRDT for merging, or split store_prefs into `remote_store_prefs` and `central_store_prefs` as a simpler version of broken up prefs.
 
I'm personally, trying to avoid having lots of different prefs in the database to need more complex querying, mental overhead to decide is this a new pref row, is it part of another pref, etc. To me the simplicity of the big struct is that everything is there in one place. It's easy to follow through the flow from db records to what gets passed to in the UI.

I might be on a different page to others though so wanted to share that rust example to explain.
I agreed there might be limitations and challenges that come up with this approach, but I think it's a good way to start and only with good reason e.g requirements & use case do we extend it further.

Each pref would record if it's relevent to apply at a global, store, or other level.

In sync we'd sync a `preference_data` record for each store and one global (for now) could add user, or machine etc as needed in future. This would have a json payload similar to Option 1.

There's nothing to stop us still deserialise them in a rust struct as needed

This option would allow us to develop a preference editor UI similar to what we have with editing store/name properties.

If you need some more complex type in your sync record, then it probably is worth creating a dedicated ui and management for this struct.

_Pros:_

- Re-uses an existing pattern
- Clear path for creating a UI without lots of over head for adding a new pref
- Can still deserialise to specific structs if needed
- Could be used by plugins to add additional prefs as needed

_Cons:_

- Could be larger payload as prefs aren't broken down into different areas
- Need to insert prefs into database rather than just adding to code? Maybe there's a code first way to achieve the same thing? Enums?

## Decision

> TBD

Reviewer: ```suggestion
Option 1
```

The list of available prefs will be defined by a struct in the Rust backend and available in graphql.
The UI will be manually created for each new pref as added to the struct.
This approach allows us to start small, with just Booleans for Global and Store Prefs initially, but can be expanded to more complex scenarios as required.

### Consequences

-  New prefs will required additional configuration UI to be manually created. But it should be a simple copy and paste type job
- There's still some potential risk for incompatible changes to come via sync to an old version for example. We'll need to be vigilant in code reviews to avoid this issue, and potentially favour creating new prefs if say a boolean pref needs to become a number for example.
